### PR TITLE
Optimize bot.trade()

### PR DESCRIPTION
**Pull request by ImHarvol**
* Check that the bot has enough item to fullfill the trade count

* Adds the option to use villager.trade(...) instead of bot.trade(villager, ...)

* Instead of putting as many items as the bot needs for 1 trade, it puts as many items as it needs for all trades (count) to be done, this way we avoid the bot putting back in the inventory the items it's going to need for the next trades (avoids some clicks)

### CODE REVIEW

File: lib/plugins/villager.js [hunk 1]
```diff
@@ -116,6 +120,10 @@ function inject (bot, { version }) {
     count = count || Trade.maximumNbTradeUses - Trade.nbTradeUses
     assert.ok(Trade.maximumNbTradeUses - Trade.nbTradeUses > 0, 'trade blocked')
     assert.ok(Trade.maximumNbTradeUses - Trade.nbTradeUses >= count)
+    if (villager.window.count(Trade.inputItem1.type) < Trade.inputItem1.count * count ||
+    (Trade.inputItem2 && villager.window.count(Trade.inputItem2.type) < Trade.inputItem2.count * count)) {
```
Comment (wvffle) [inline lib/plugins/villager.js:124]: Uh. That if is awful. Move some stuff/conditions into variables

File: lib/plugins/villager.js [hunk 2]
```diff
@@ -150,12 +170,22 @@ function inject (bot, { version }) {
...
   }
 
-  function putRequirements (window, Trade, cb) {
-    deposit(window, Trade.inputItem1.type, Trade.inputItem1.metadata, Trade.inputItem1.count, 0, (err) => {
+  function putRequirements (window, Trade, count, cb) {
+    let input1 = 0
+    if (!window.slots[0] || window.slots[0].count < Trade.inputItem1.count) {
+      input1 = Math.min(Trade.inputItem1.stackSize, Trade.inputItem1.count * count)
+      if (window.slots[0]) input1 -= window.slots[0].count
+    }
```
Comment (wvffle) [inline lib/plugins/villager.js:178]: I think that this can be simplified and it for sure can look better

File: lib/plugins/villager.js [hunk 3]
```diff
@@ -150,12 +170,22 @@ function inject (bot, { version }) {
...
+    deposit(window, Trade.inputItem1.type, Trade.inputItem1.metadata, input1, 0, (err) => {
       if (err) {
         cb(err)
       } else if (Trade.hasItem2) {
-        deposit(window, Trade.inputItem2.type, Trade.inputItem2.metadata, Trade.inputItem2.count, 1, (err) => {
+        let input2 = 0
+        if (!window.slots[1] || window.slots[1].count < Trade.inputItem2.count) {
+          input2 = Math.min(Trade.inputItem2.stackSize, Trade.inputItem2.count * count)
+          if (window.slots[1]) input2 -= window.slots[1].count
+        }
```
Comment (wvffle) [inline lib/plugins/villager.js:187]: Same here, it can be a lot more readable

**Review by wvffle — CHANGES_REQUESTED** [review]

File: lib/plugins/villager.js (hunk 2 above)
Comment (ImHarvol) [reply lib/plugins/villager.js:178]: Is this what you mean?
```js
    let input1 = 0
    if (!window.slots[0]) {
      input1 = Math.min(Trade.inputItem1.stackSize, Trade.inputItem1.count * count)
    } else if (window.slots[0].count < Trade.inputItem1.count) {
      input1 = Math.min(Trade.inputItem1.stackSize, Trade.inputItem1.count * count) - window.slots[0].count
    }
```

Comment (wvffle) [reply lib/plugins/villager.js:178]: Yeah, it's a bit more readable, you could even do like that or something similar. What I mean is that there's just too much similar long variables that makes this part of code hard to read
```js
const [slot] = window.slots
const { count: tradeCount1, stackSize: stackSize1 } = Trade.inputItem1
const realCount1 = Math.min(stackSize1, tradeCount1 * count)

const input1 = !slot 
  ? realCount1
  : (slot.count < tradeCount1 ? realCount1 - slot.count : 0)
```

Comment (ImHarvol) [reply lib/plugins/villager.js]: Yeah that's much clear

File: lib/plugins/villager.js [hunk 4]
```diff
@@ -138,9 +147,26 @@ function inject (bot, { version }) {
...
-                villager.window.updateSlot(1, null)
+                const slot1 = villager.window.slots[0]
+                if (slot1) {
+                  assert.strictEqual(slot1.type, Trade.inputItem1.type)
+                  const updatedCount1 = slot1.count - Trade.inputItem1.count
+                  const updatedSlot1 = updatedCount1 <= 0
+                    ? null
+                    : { ...slot1, count: updatedCount1 }
+                  villager.window.updateSlot(0, updatedSlot1)
+                }
```
Comment (wvffle) [inline lib/plugins/villager.js:158]: I'd just add a \n here

File: lib/plugins/villager.js [hunk 5]
```diff
@@ -138,9 +147,26 @@ function inject (bot, { version }) {
...
+                const slot2 = villager.window.slots[1]
```
Comment (wvffle) [inline lib/plugins/villager.js:159]: I'd do `const [slot1, slot2] = villager.window.slots`

File: lib/plugins/villager.js [hunk 6]
```diff
@@ -150,12 +176,26 @@ function inject (bot, { version }) {
     }
   }
 
-  function putRequirements (window, Trade, cb) {
-    deposit(window, Trade.inputItem1.type, Trade.inputItem1.metadata, Trade.inputItem1.count, 0, (err) => {
+  function putRequirements (window, Trade, count, cb) {
+    const slot1 = window.slots[0]
```
Comment (wvffle) [inline lib/plugins/villager.js:180]: Same here, `const [slot1, slot2] = window.slots`

File: lib/plugins/villager.js [hunk 7]
```diff
@@ -150,12 +176,26 @@ function inject (bot, { version }) {
...
+    const { count: tradeCount1, stackSize: stackSize1 } = Trade.inputItem1
+    const needCount1 = Math.min(stackSize1, tradeCount1 * count)
```
Comment (wvffle) [inline lib/plugins/villager.js:182]: neededCount

File: lib/plugins/villager.js [hunk 8]
```diff
@@ -150,12 +176,26 @@ function inject (bot, { version }) {
...
+
+    const input1 = !slot1
+      ? needCount1
+      : (slot1.count < tradeCount1 ? needCount1 - slot1.count : 0)
+    deposit(window, Trade.inputItem1.type, Trade.inputItem1.metadata, input1, 0, (err) => {
```
Comment (wvffle) [inline lib/plugins/villager.js:187]: I'd move type and metadata to the destruction as `type1` and `metadata1`

**Review by wvffle — APPROVED** [review]
Overall LGTM but personally I'd make some changes

**Comment by ImHarvol** [comment]
Everything makes sense, thank you for the good practices!

File: lib/plugins/villager.js [hunk 9]
```diff
@@ -117,6 +121,17 @@ function inject (bot, { version }) {
     assert.ok(Trade.maximumNbTradeUses - Trade.nbTradeUses > 0, 'trade blocked')
     assert.ok(Trade.maximumNbTradeUses - Trade.nbTradeUses >= count)
 
+    const itemCount1 = villager.window.count(Trade.inputItem1.type, Trade.inputItem1.metadata)
+    const hasEnoughItem1 = itemCount1 >= Trade.inputItem1.count * count
+    let hasEnoughItem2 = true
+    if (Trade.inputItem2) {
+      const itemCount2 = villager.window.count(Trade.inputItem2.type, Trade.inputItem2.metadata)
+      hasEnoughItem2 = itemCount2 >= Trade.inputItem2.count * count
+    }
```
Comment (wvffle) [inline lib/plugins/villager.js:130]: And here, I'd add some spacing around this if

**Comment by wvffle** [comment]
Also, can  you write tests for your function?

**Review by rom1504 — APPROVED** [review]

**Pull request merged**